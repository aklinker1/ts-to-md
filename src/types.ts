import * as prettier from 'prettier';

/**
 * Options for customizing the generated markdown.
 */
export interface Options {
  /**
   * Path to your Typescript project (folder with `tsconfig.json` or path to `tsconfig.json`)
   */
  project?: string;
  /**
   * The `.ts` or `.d.ts` file you want to generate documentation for.
   */
  inputFile: string;
  /**
   * A template for displaying the type definition. To hide the definition, set to an empty string.
   *
   * @default "```ts\n// Definition\n{{template}}\n```"
   */
  definitionTemplate?: string;
  /**
   * When true, don't throw an error when there is a type erorr.
   */
  skipTypeCheck?: boolean;
  /**
   * Config for prettier, or `false` to disable formatting. Default to the config found in your
   * `process.cwd()`
   */
  prettier?: false | prettier.Config;
  /**
   * Set to true to prevent a table of contents from being generated.
   */
  disableToc?: boolean;
}

export interface Ctx {
  definitions: Record<string, boolean>;
}
