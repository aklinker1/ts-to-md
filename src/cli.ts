#!/usr/bin/env node
import meow from 'meow';
import path from 'node:path';
import fs from 'node:fs';
import { generate } from '.';

const cli = meow(
  `
	Usage
	  $ ts-to-js <input> <output>

	Options
	  --project, -p           Path to tsconfig.json to use.
	  --definitionTemplate    A string used to change what the TS definition looks like. Set to "" to exclude TS definitions.
    --skipTypeCheck, -s     Skip checking typescript for type errors.
  `,
  {
    importMeta: import.meta,
    flags: {
      project: {
        type: 'string',
        alias: 'p',
        default: path.resolve('tsconfig.json'),
      },
      definitionTemplate: {
        type: 'string',
      },
      skipTypeCheck: {
        type: 'boolean',
      },
    },
  }
);

const [inputFile, outputFile] = cli.input;
console.log(`\nGenerating ${outputFile}...`);

const markdown = generate({
  inputFile,
  project: cli.flags.project,
  definitionTemplate: cli.flags.definitionTemplate,
  skipTypeCheck: cli.flags.skipTypeCheck,
});

fs.writeFileSync(outputFile, markdown, 'utf-8');
console.log('Done!');
