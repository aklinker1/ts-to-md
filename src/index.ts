import {
  ExportedDeclarations,
  FunctionDeclaration,
  JSDoc,
  Node,
  ParameterDeclaration,
  Project,
  PropertySignature,
  SyntaxKind,
} from 'ts-morph';
import * as prettier from 'prettier';
import 'prettier/parser-markdown.js';
import { getPublicDefinitions } from './utils/getPublicDefinitions';
import { Ctx, Options } from './types';
import { getMarkdownTypeRef } from './utils/getMarkdownTypeRef';

export { Options };

/**
 * Generate and return the documentation for a TS file as a string.
 *
 * This method does not write the config to a file, you'll need to do that yourself.
 */
export function generate(options: Options): string {
  const project = getProject(options);
  const file = project.getSourceFileOrThrow(options.inputFile);

  const definitions = getPublicDefinitions(file);

  const ctx: Ctx = {
    definitions: definitions.reduce<Ctx['definitions']>((map, def) => {
      map[def.name] = true;
      return map;
    }, {}),
  };

  const blocks = definitions.map(({ name, def }) =>
    getDefinitionMarkdown({ ctx, name, def, options })
  );

  const lines = ['# API'];
  lines.push(...blocks);

  let markdown = lines.join('\n\n');
  if (options.prettier !== false) {
    markdown = prettier.format(markdown, {
      ...options.prettier,
      parser: 'markdown',
    });
  }
  return markdown;
}

function getProject(options: Options): Project {
  const project = new Project({
    tsConfigFilePath: options.project,
    // skipAddingFilesFromTsConfig: true,
  });
  project.addSourceFilesAtPaths(options.inputFile);

  if (!options.skipTypeCheck) {
    const errors = project.getPreEmitDiagnostics();
    if (errors.length > 0) {
      console.error(project.formatDiagnosticsWithColorAndContext(errors));
      throw Error('There are type errors. Markdown will not be generated.');
    }
  }

  return project;
}

function getDefinitionMarkdown({
  ctx,
  name,
  def,
  options,
}: {
  ctx: Ctx;
  name: string;
  def: ExportedDeclarations;
  options: Options;
}): string {
  const lines = [`## \`${name}\``, getTypeDefMarkdownCodeblock(options, def)];

  const description = getJsDocNode(def)?.getDescription();
  if (description) lines.push(description);

  if (def.isKind(SyntaxKind.FunctionDeclaration)) {
    const returnType = getMarkdownTypeRef(ctx, def.getReturnTypeNode());
    if (returnType !== '`void`') lines.push(`Returns ${returnType}.`);

    const params = def.getParameters();
    if (params.length > 0)
      lines.push(getParamsMarkdownTable(ctx, { fn: def, params }));
  } else if (def.isKind(SyntaxKind.InterfaceDeclaration)) {
    const properties = def.getProperties();
    if (properties.length > 0)
      lines.push(getPropertiesMarkdownTable(ctx, properties));
  }

  return lines.join('\n\n');
}

function getParamsMarkdownTable(
  ctx: Ctx,
  { fn, params }: { fn: FunctionDeclaration; params: ParameterDeclaration[] }
): string {
  const lines = [
    `| Parameter | Type | Optional | Default | Description |`,
    `| --------- | ---- | :------: | ------- | ----------- |`,
  ];
  params.forEach((param) => {
    const name = `\`${param.getName()}\``;
    const initializer = param.getInitializer();
    const optional = param.isOptional() || initializer ? '✅' : '';
    const defaultValue = initializer
      ? getMarkdownTypeRef(ctx, initializer.getType())
      : '';
    const type = getMarkdownTypeRef(
      ctx,
      initializer
        ? initializer.getType().getApparentType()
        : param.getTypeNode()
    );
    const tags = getJsDocNode(fn)?.getTags();
    const tag = tags?.find((tag) =>
      tag
        .getText()
        .replace('@param', '')
        .trimStart()
        .startsWith(param.getName())
    );
    const comment = tag?.getCommentText()?.replaceAll('\n', ' ') ?? '';
    const description = comment.replace(/^\s*[:-]\s*/, '');
    lines.push(
      `| ${name} | ${type} | ${optional} | ${defaultValue} | ${description}`
    );
  });
  return lines.join('\n');
}

function getPropertiesMarkdownTable(
  ctx: Ctx,
  properties: PropertySignature[]
): string {
  const lines = [
    `| Field | Type | Optional | Description |`,
    `| ----- | ---- | :------: | ----------- |`,
  ];
  properties.forEach((property) => {
    const name = `\`${property.getName()}\``;
    const optional = property.hasQuestionToken() ? '✅' : '';
    const type = getMarkdownTypeRef(ctx, property.getTypeNode());
    const description =
      getJsDocNode(property)?.getCommentText()?.replaceAll('\n', ' ') ?? '';
    lines.push(`| ${name} | ${type} | ${optional} | ${description}`);
  });
  return lines.join('\n');
}

function getTypeDefMarkdownCodeblock(
  options: Options,
  def: ExportedDeclarations
) {
  const template =
    options.definitionTemplate ?? '```ts\n// Definition\n{{template}}\n```';

  if (def.isKind(SyntaxKind.FunctionDeclaration)) def.removeBody();
  const defStr = def.print({ removeComments: true });

  return template.replace('{{template}}', defStr);
}

function getJsDocNode(def: Node): JSDoc | undefined {
  if (Node.isJSDocable(def)) return def.getJsDocs()[0];
  return undefined;
}
