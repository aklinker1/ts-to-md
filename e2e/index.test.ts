import { describe, it, expect } from 'vitest';
import fs from 'node:fs';
import path from 'node:path';
import { generate } from '../src';

const dir = path.resolve('e2e/cases');
const testCases = fs.readdirSync(dir).filter((file) => file.endsWith('.ts'));

describe('TypeScript to Markdown Generation', () => {
  it.each(testCases)('should convert %s to markdown', (inputFileName) => {
    const inputFile = path.join(dir, inputFileName);
    const outputFile = inputFile.replace('.ts', '.md');

    const actual = generate({
      inputFile: path.join(dir, inputFileName),
      project: path.resolve('tsconfig.json'),
      prettier: { semi: false, printWidth: 120 },
    });

    // Write output to file for visual inspection.
    fs.writeFileSync(outputFile, actual, 'utf-8');

    expect(actual).toMatchSnapshot();
  });
});
