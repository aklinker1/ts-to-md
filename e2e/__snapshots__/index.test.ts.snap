// Vitest Snapshot v1

exports[`TypeScript to Markdown Generation > should convert basic-references.ts to markdown 1`] = `
"# API

- <code><a href=\\"#defineconfig\\">defineConfig</a></code>
- <code><a href=\\"#userconfig\\">UserConfig</a></code>

## <code>defineConfig</code>

\`\`\`ts
// Definition
export function defineConfig(config: UserConfig): UserConfig
\`\`\`

An identity function used to ensure type safety when defining config.

Returns <code><a href=\\"#userconfig\\">UserConfig</a></code>.

| Parameter | Type                                              | Optional | Default | Description |
| --------- | ------------------------------------------------- | :------: | ------- | ----------- |
| \`config\`  | <code><a href=\\"#userconfig\\">UserConfig</a></code> |          |         |

## <code>UserConfig</code>

\`\`\`ts
// Definition
export interface UserConfig {
  name: string
}
\`\`\`

Configuration provided by the user

| Field  | Type                | Optional | Description |
| ------ | ------------------- | :------: | ----------- |
| \`name\` | <code>string</code> |          |
"
`;

exports[`TypeScript to Markdown Generation > should convert constant.ts to markdown 1`] = `
"# API

- <code><a href=\\"#some_value\\">SOME_VALUE</a></code>

## <code>SOME_VALUE</code>

\`\`\`ts
// Definition
SOME_VALUE = 123
\`\`\`
"
`;

exports[`TypeScript to Markdown Generation > should convert empty.ts to markdown 1`] = `
"# API
"
`;

exports[`TypeScript to Markdown Generation > should convert function.ts to markdown 1`] = `
"# API

- <code><a href=\\"#somefunction\\">someFunction</a></code>

## <code>someFunction</code>

\`\`\`ts
// Definition
export async function someFunction(arg1: string, arg2?: number, arg3 = true): Promise<string>
\`\`\`

Some function docs.

Returns <code>Promise&lt;string&gt;</code>.

| Parameter | Type                 | Optional | Default | Description         |
| --------- | -------------------- | :------: | ------- | ------------------- |
| \`arg1\`    | <code>string</code>  |          |         | The first argument  |
| \`arg2\`    | <code>number</code>  |    ✅    |         | The second argument |
| \`arg3\`    | <code>boolean</code> |    ✅    | true    | The third argument  |
"
`;

exports[`TypeScript to Markdown Generation > should convert interface.ts to markdown 1`] = `
"# API

- <code><a href=\\"#test\\">Test</a></code>

## <code>Test</code>

\`\`\`ts
// Definition
export interface Test {
  field1: boolean
  field2?: number
  field3: string | undefined
}
\`\`\`

Test interface description

| Field    | Type                                 | Optional | Description  |
| -------- | ------------------------------------ | :------: | ------------ |
| \`field1\` | <code>boolean</code>                 |          | Field 1 docs |
| \`field2\` | <code>number</code>                  |    ✅    |
| \`field3\` | <code>string &#124; undefined</code> |          | Field 3 docs |
"
`;

exports[`TypeScript to Markdown Generation > should convert serialize-types.ts to markdown 1`] = `
"# API

- <code><a href=\\"#alltypes\\">AllTypes</a></code>

## <code>AllTypes</code>

\`\`\`ts
// Definition
export interface AllTypes {
  void: void
  never: never
  undefined: undefined
  null: null
  any: any
  object: {
    key: \\"value\\"
  }
  array: string[]
  readonlyArray: ReadonlyArray<string>
  union: \\"one\\" | \\"two\\"
  intersection: \\"one\\" & \\"two\\"
  string: string
  stringLiteral: \\"hello\\"
  number: number
  numberLiteral: 44
  boolean: boolean
  booleanLiteral: false
  anonymousTuple: [string, number]
  namedTuple: [one: string, two: number]
  customType: CustomType
  singleTypeParam: Promise<any>
  multipleTypeParams: Pick<
    {
      one: any
      two: any
    },
    \\"one\\"
  >
}
\`\`\`

| Field                | Type                                     | Optional | Description |
| -------------------- | ---------------------------------------- | :------: | ----------- |
| \`void\`               | <code>void</code>                        |          |
| \`never\`              | <code>never</code>                       |          |
| \`undefined\`          | <code>undefined</code>                   |          |
| \`null\`               | <code>null</code>                        |          |
| \`any\`                | <code>any</code>                         |          |
| \`object\`             | <code>object</code>                      |          |
| \`array\`              | <code>Array&lt;string&gt;</code>         |          |
| \`readonlyArray\`      | <code>ReadonlyArray&lt;string&gt;</code> |          |
| \`union\`              | <code>\\"one\\" &#124; \\"two\\"</code>          |          |
| \`intersection\`       | <code>never</code>                       |          |
| \`string\`             | <code>string</code>                      |          |
| \`stringLiteral\`      | <code>\\"hello\\"</code>                     |          |
| \`number\`             | <code>number</code>                      |          |
| \`numberLiteral\`      | <code>44</code>                          |          |
| \`boolean\`            | <code>boolean</code>                     |          |
| \`booleanLiteral\`     | <code>false</code>                       |          |
| \`anonymousTuple\`     | <code>[string, number]</code>            |          |
| \`namedTuple\`         | <code>[string, number]</code>            |          |
| \`customType\`         | <code>CustomType</code>                  |          |
| \`singleTypeParam\`    | <code>Promise&lt;any&gt;</code>          |          |
| \`multipleTypeParams\` | <code>Pick&lt;object, \\"one\\"&gt;</code>   |          |
"
`;

exports[`TypeScript to Markdown Generation > should convert sorting.ts to markdown 1`] = `
"# API

- <code><a href=\\"#a\\">a</a></code>
- <code><a href=\\"#a\\">A</a></code>
- <code><a href=\\"#b\\">b</a></code>
- <code><a href=\\"#b\\">B</a></code>
- <code><a href=\\"#c\\">c</a></code>
- <code><a href=\\"#c\\">C</a></code>

## <code>a</code>

\`\`\`ts
// Definition
a = 0
\`\`\`

## <code>A</code>

\`\`\`ts
// Definition
A = 3
\`\`\`

## <code>b</code>

\`\`\`ts
// Definition
b = -1
\`\`\`

## <code>B</code>

\`\`\`ts
// Definition
B = 2
\`\`\`

## <code>c</code>

\`\`\`ts
// Definition
c = -1
\`\`\`

## <code>C</code>

\`\`\`ts
// Definition
C = 1
\`\`\`
"
`;
